Query Refactoring: Moving parser NAME constant to corresponding query builder

cbuescher opened the pull request:
As part of the query refactoring, this PR moves the NAME field that currently identifies each query parser to the corresponding query builder. The reason for this, as discussed in https://github.com/elastic/elasticsearch/pull/11121#discussion_r30233757, is that we need still need to be able to link parser and builder implementations, but that the query builder is independent of the parser (queries don't necessarily need to be coverted to XContent any more). Builders don't need to know about their parsers, but parsers need to be linked to their respective builders.

This change does the following:
- move the NAME constants from the parsers to the builders, making sure they are unique
- rename the existing getParserName() in the Builder classes to queryId(), returning the unique NAME constant. This is now public so we can retrieve query name when coding against the QueryBuilder abstract class
-  modify the names() method in the parsers so it contains at least one member that is the query id

cbuescher committed:
Query Refactoring: Moving parser NAME constant to corresponding query builder

As part of the query refactoring, this PR moves the NAME field that currently identifies each query parser to the
corresponding query builder. The reason for this, as discussed in #11121 (comment), is that we need still need to
be able to link parser and builder implementations, but that the query builder is independent of the parser (queries
don't necessarily need to be coverted to XContent any more). Builders don't need to know about their parsers, but
parsers need to be linked to their respective builders.

Closes #11178

cbuescher commented:
As a further explanation: there were two cases where the 1:1 mapping from parsers to builders was not completely clear:
* FQueryFilterParser didn't have a corresponding Builder class, instead queries with the `"fquery"` id were build in the (already deprecated) QueryFilterBuilder. To make this link clearer I introduced a new FQueryFilterParser that extends QueryFilterBuilder but otherwise shares the existing code there, just carries the moved NAME constant
* TermsLookupQueryBuilder was a builder without a correspondingly names parser. It used the TermsQueryParser and so also its NAME. To have a unique id for each builder I introduced the new name `
`"terms_lookup"` here and added this to the names() array in the TermsQueryParser. Because all these names are used to register the parsers this will link the builder to that parser now.

javanna commented on src/main/java/org/elasticsearch/index/query/BoolQueryBuilder.java:
```diff
@@ -200,7 +202,7 @@ private void doXArrayContent(String field, List<QueryBuilder> clauses, XContentB
-    protected String parserName() {
-        return BoolQueryParser.NAME;
+    public String queryId() {
+        return BoolQueryBuilder.NAME;
```
in all the builders, you can remove the class name, just `return NAME` is enough.

javanna commented on src/main/java/org/elasticsearch/index/query/QueryBuilder.java:
```diff
@@ -45,6 +45,11 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
     /**
+     * @return a unique name this query is identified with
+     */
+    abstract public String queryId();
```
I have a small preference over public abstract rather than abstract public

nik9000 replied:
There is actually a [standard](http://checkstyle.sourceforge.net/config_modifier.html) for this if you particularly enjoy standards.

javanna replied:
yea thanks for the link, correct order is public abstract, which I think we've been using in the rest of the codebase.

javanna commented on src/main/java/org/elasticsearch/index/query/TermsQueryParser.java:
```diff
@@ -62,7 +61,7 @@ public TermsQueryParser() {
     @Override
     public String[] names() {
-        return new String[]{NAME, "in"};
+        return new String[]{TermsQueryBuilder.NAME, TermsLookupQueryBuilder.NAME, "in"};
```
I'd rather merge TermsQueryBuilder and TermsLookupBuilder, this is going to be a problem anyway when registering builders for serialization if we want to keep registering the parser only and deduce the builder from it . I don't see any value in having two builders for the same query, it becomes confusing for java api users too.

Also, the easy way of creating queries should be through querybuilders, we can keep the existing `QueryBuilders#termsLookupQuery` but it will return a TermsQueryBuilder instead. We can also add more methods to `QueryBuilders` to create a terms lookup query that holds all the needed params.

javanna replied:
I don't think you have to add here `TermsLookupQueryBuilder.NAME` to the array

javanna commented on src/main/java/org/elasticsearch/index/query/FQueryFilterBuilder.java:
```diff
@@ -0,0 +1,38 @@
+    public FQueryFilterBuilder(QueryBuilder queryBuilder) {
+        super(queryBuilder);
+    }
+
```
The QueryFilterBuilder is going to be interesting when we will try and address #10776 as it doesn't support query name actually but f query filter does.... this is confusing to me but we can't do much about it right now I guess. I think I would do a proper separation of FQueryFilterBuilder and QueryFilterBuilder though. let's have this 1:1 with the rest layer, that means moving the query name here and splitting the current toXContent into two, one that generates query filter and one to be moved here that generates fquery.

javanna commented:
good change! left a few comments

cbuescher commented:
I rebased this PR after recent changes on feature branch and addressed the two comments above. The two main changes here:
* TermsQueryBuilder covers all of former TermsLookupBuilder functionality, leaving deprecated Stub class, leading to a 1:1 relation between Builder and Parser here.
* FQueryFilterBuilder directly extends QueryBuilder, moved cases where queryName is set out of QueryFilterBuilder.

javanna commented on src/main/java/org/elasticsearch/index/query/QueryFilterBuilder.java:
```diff
@@ -29,47 +29,29 @@
-                builder.field("_name", queryName);
-            }
-            builder.endObject();
-        }
```
I thought we discussed (missing bits in the review comments) to leave this as-is and to not introduce `FQueryFilterBuilder` as it is not needed. This change breaks bw compatibility as you pointed out, thus we should keep this class as-is, and if we do that it makes little sense to introduce a deprecated additional `FQueryFilterBuilder`. The idea is that two parsers sharing the same builder should not be a problem....unless I am missing something.

cbuescher replied:
All right, will remove FQueryFilterBuilder then, put the NAME of the extra FQueryFilterParser somewhere else in the QueryFilterBuilder then.

javanna commented on src/test/java/org/elasticsearch/transport/ContextAndHeaderTransportTests.java:
```diff
@@ -122,7 +122,7 @@ public void testThatTermsLookupGetRequestContainsContextAndHeaders() throws Exce
         transportClient().admin().indices().prepareRefresh(queryIndex, lookupIndex).get();
 
-        TermsLookupQueryBuilder termsLookupFilterBuilder = QueryBuilders.termsLookupQuery("username").lookupIndex(lookupIndex).lookupType("type").lookupId("1").lookupPath("followers");
+        TermsQueryBuilder termsLookupFilterBuilder = QueryBuilders.termsLookupQuery("username", lookupIndex, "type", "1", "followers");
```
I wonder why all these calls changed, it was not needed right? just your personal preference? Double checking that we are bw compatible.

cbuescher replied:
I probably went to far with refactoring here by changing the termsLookupQuery() method signature. Will revert that again.

javanna commented:
I did another review round, thanks @cbuescher

cbuescher commented:
@javanna I had another round, mostly reverting some of the changes in my prevoius commits.

javanna commented on src/main/java/org/elasticsearch/index/query/TermsQueryBuilder.java:
```diff
@@ -131,24 +140,80 @@ public TermsQueryBuilder queryName(String queryName) {
-
         if (queryName != null) {
-            builder.field("_name", queryName);
+                        builder.field("_name", queryName);
```
too many tabs?

javanna commented:
left a super minor comment on indentation, LGTM besides that
